### PR TITLE
chore(email): Switch to \Zend\Mail library instead of raw mail() function

### LINKED ISSUE #5918: Adopt third-party email library
Our email management is scary and hacky. There are plenty of good, object-oriented solutions out there that come with attachment support and such.

* [Symfony/Swiftmailer](http://symfony.com/doc/current/cookbook/email/email.html) (MIT)
* [Zend](https://github.com/zendframework/zf2/tree/master/library/Zend/Mail) (3-Clause BSD)
* [PHPMailer](https://github.com/PHPMailer/PHPMailer) (LGPL 2.1)

Right now I'm leaning Zend because it also has support for *receiving* emails, which would be amazing.

### PR DESCRIPTION
**Pull request by ewinslow**
This lays a good foundation for improving Elgg's flexibility and security
in sending emails. Some of the features we open up because of this change:
- HTML emails
- Attachments
- Improved security (automatic header escaping, etc.)
- Third-party SMTP services (e.g. Mandrill)
- Receiving emails
- In-memory transport for easier testing/development

We aren't introducing our own interfaces because `Zend\Mail`
is plenty well supported and the API is agreeable to us.
Writing a wrapper would just be unnecessary overhead.

BREAKING CHANGE:
We are switching to `Zend\Mail` for sending emails in Elgg 2.0.
It's likely that there are some edge cases that the library
handles differently than Elgg 1.x used to. Take care to test
your email notifications carefully when upgrading to 2.0.

Fixes #5918

### CODE REVIEW

**Comment by ewinslow** [comment]
Note: this is UNTESTED atm.

**Comment by ewinslow** [comment]
When we announce this we can focus on the benefits of:
 * HTML email support
 * Attachment support
 * Improved security (automatic header escaping, etc.)
 * Performance
   * Can utilize a third party transport like Mandrill instead of localhost
   * Can load-balance transports

File: engine/lib/notification.php [hunk 1]
```diff
@@ -623,62 +642,25 @@ function elgg_send_email($from, $to, $subject, $body, array $params = null) {
...
-	// Sanitise subject by stripping line endings
-	$subject = preg_replace("/(\r\n|\r|\n)/", " ", $subject);
-	// this is because Elgg encodes everything and matches what is done with body
-	$subject = html_entity_decode($subject, ENT_QUOTES, 'UTF-8'); // Decode any html entities
-	if (is_callable('mb_encode_mimeheader')) {
-		$subject = mb_encode_mimeheader($subject, "UTF-8", "B");
-	}
-
-	// Format message
-	$body = html_entity_decode($body, ENT_QUOTES, 'UTF-8'); // Decode any html entities
```
Comment (mrclay) [inline engine/lib/notification.php]: Unless we're switching all to html mail, I think we need these lines.

Comment (mrclay) [inline engine/lib/notification.php]: Particularly HTML decode followed by stripping tags. Subject needs the same.

**Comment by mrclay** [comment]
Strawman based on my "env" service PR: in settings.php allow `$CONFIG->mailTransportFactory` to be a Closure that returns a valid transport.

**Comment by mrclay** [comment]
Actually I wonder if we need to reserve `$CONFIG->factories` as an array of Closures for injecting services (but we only handle those we support, not arbitrary keys).

**Comment by ewinslow** [comment]
Originally I had considered zend's library. I'm still open to that. The attachments api isn't super nice but they have support for receiving mail which seems killer.

File: engine/lib/notification.php [hunk 2]
```diff
@@ -577,6 +577,35 @@ function set_user_notification_setting($user_guid, $method, $value) {
...
+ * @param string $contact e.g. "Evan <evan@elgg.org>"
+ * 
+ * @return array e.g. ['name' => 'Evan', 'email' => 'evan@elgg.org']
+ * @access private
+ */
+function _elgg_parse_email_contact($contact) {
+	$containsName = preg_match('/<(.*)>/', $contact, $matches) == 1;
+	if ($containsName) {
+		return [
+			'name' => trim(elgg_substr($contact, 0, $email_pos)),
```
Comment (mrclay) [inline engine/lib/notification.php]: Where does $email_pos come from? If $containsName, just use the same pattern with preg_replace to remove it and trim the result?

File: engine/lib/notification.php [hunk 3]
```diff
@@ -593,16 +622,6 @@ function set_user_notification_setting($user_guid, $method, $value) {
 function elgg_send_email($from, $to, $subject, $body, array $params = null) {
 	global $CONFIG;
 
-	if (!$from) {
```
Comment (mrclay) [inline engine/lib/notification.php]: Why are these checks going away?

File: engine/lib/notification.php (hunk 2 above)
Comment (ewinslow) [inline engine/lib/notification.php]: Should be `elgg_strpos($contact, '<')`

File: engine/lib/notification.php (hunk 3 above)
Comment (ewinslow) [inline engine/lib/notification.php]: Over eager cleanup

Comment (ewinslow) [inline engine/lib/notification.php]: Will restore

**Comment by mrclay** [comment]
Obviously needs mention in docs/guides/upgrading

**Comment by beck24** [comment]
The switch to SwiftMailer is good, should we also default to html emails?  We're at a point where every email client handles them just fine and *every* project I work on requires html emails as a feature.

**Comment by ewinslow** [comment]
Switching to HTML emails sounds great and this will make that possible without hacks, but I think that can/should come as a separate effort.

**Comment by mrclay** [comment]
...As in get this well integrated, then attack HTML in a separate PR. Sounds good.

**Comment by ewinslow** [comment]
More reasons I've reverted to my preference for Zend's Mail library, in addition to the ability to read/receive emails.

 * In-memory transport which can be used during testing/development.
 * Provides a ridiculously simple mailer/transport interface: a single `send(Message $message)` method

**Comment by ewinslow** [comment]
OK, I believe I addressed all questions with this update. Some changes I made:

 * Switched to Zend, since it offers more features and has a cleaner API, in my estimation.
 * Moved new private helper functions to static methods on a new `Elgg\Mail\Address` helper class.
 * Added a test for this function (it's actually possible now!)
 * Added docs to upgrading guide
 * Restored previously deleted functionality to elgg_send_email

File: engine/lib/notification.php [hunk 4]
```diff
@@ -623,62 +628,33 @@ function elgg_send_email($from, $to, $subject, $body, array $params = null) {
...
-		$subject = mb_encode_mimeheader($subject, "UTF-8", "B");
-	}
-
-	// Format message
-	$body = html_entity_decode($body, ENT_QUOTES, 'UTF-8'); // Decode any html entities
+	$body = html_entity_decode($result['body'], ENT_QUOTES, 'UTF-8');
 	$body = elgg_strip_tags($body); // Strip tags from message
-	$body = preg_replace("/(\r\n|\r)/", "\n", $body); // Convert to unix line endings in body
-	$body = preg_replace("/^From/", ">From", $body); // Change lines starting with From to >From
 	$body = wordwrap($body);
```
Comment (mrclay) [inline engine/lib/notification.php]: Does Zend not do this automatically?

File: engine/tests/phpunit/Elgg/Mail/MailerTest.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?php
+namespace Elgg\Mail;
+
+use PHPUnit_Framework_TestCase as TestCase;
+use Zend\Mail\Transport\InMemory as InMemoryTransport;
+
+class MailerTest extends TestCase {
+	
+	function testElggSendEmailPassesAllFieldsAsMessageToMailer() {
+		$mailer = new InMemoryTransport();
```
Comment (mrclay) [inline engine/tests/phpunit/Elgg/Mail/MailerTest.php]: yeah, this is awesome

Comment (mrclay) [inline engine/tests/phpunit/Elgg/Mail/MailerTest.php]: We should inject this in the PHPUnit bootstrap

File: engine/lib/notification.php [hunk 6]
```diff
@@ -623,62 +628,33 @@ function elgg_send_email($from, $to, $subject, $body, array $params = null) {
...
-
-	return mail($to, $subject, $body, $headers_string);
+	
+	$message = new Message();
+	$message->addFrom($from_address);
```
Comment (mrclay) [inline engine/lib/notification.php]: need to set encoding

**Comment by ewinslow** [comment]
@mrclay Can you submit a PR to my PR, or perhaps just LGTM this and follow up with another one later to fix these bugs? They both seem like good ideas but not blockers, and I'm a little busy these days with the :baby:

**Comment by mrclay** [comment]
I sent you https://github.com/ewinslow/Elgg/pull/6

**Comment by juho-jaakkola** [comment]
I don't have time to review right now, but I'm definitely not blocking.

**Comment by mrclay** [comment]
I'm going to make a new PR with this + https://github.com/ewinslow/Elgg/pull/6.

**Comment by mrclay** [comment]
Replaced by #8331
